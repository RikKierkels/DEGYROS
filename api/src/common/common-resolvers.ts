import { DateTimeResolver } from 'graphql-scalars';
import { Resolvers } from '../generated/graphql';

const resolvers: Resolvers = {
  DateTime: DateTimeResolver,
};

export default resolvers;
