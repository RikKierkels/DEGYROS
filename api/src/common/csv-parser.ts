import { parse } from 'papaparse';

export const parseCsv = <T>(headers: string[], transform: (value: string) => any) => (text: string) =>
  parse<T>(text, {
    header: true,
    dynamicTyping: true,
    skipEmptyLines: true,
    transform,
    transformHeader(_, index: number): string {
      return headers[index];
    },
  });
