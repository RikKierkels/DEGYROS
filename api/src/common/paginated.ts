import { PageInput, ResolversParentTypes } from '../generated/graphql';

type Pageables<T> = { [K in keyof T]: T[K] extends { items: unknown } ? T[K] : never }[keyof T];
type PageableItems<T> = T extends { items: infer Items } ? (Items extends Array<infer Item> ? Item : never) : never;

type PageableResolverTypes = Pageables<ResolversParentTypes>;
type PageableResolverItems = PageableItems<PageableResolverTypes>;

const paginated = <Item extends PageableResolverItems>(
  items: Item[],
  page: PageInput,
  total: number,
): PageableResolverTypes => ({
  items,
  total,
  count: items.length,
  size: page.size || items.length,
  offset: page.offset,
});

export default paginated;
