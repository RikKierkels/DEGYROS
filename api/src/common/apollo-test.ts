import { createApolloServer } from '../apollo';
import { createTestClient } from 'apollo-server-testing';

export const createApolloTestClient = () => {
  const server = createApolloServer();
  return createTestClient(server as any);
};
