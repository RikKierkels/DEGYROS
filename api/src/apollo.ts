import { ApolloServer, makeExecutableSchema } from 'apollo-server';
import { MongoClient } from 'mongodb';
import { join } from 'path';
import { DIRECTIVES as typeDefsMongo } from '@graphql-codegen/typescript-mongodb';
import { loadFilesSync } from '@graphql-tools/load-files';
import typeDefs from './schema';
import { MongoDataSource } from './common/mongo-datasource';
import { TransactionDbObject } from './generated/graphql';

export type TransactionsDb = MongoDataSource<TransactionDbObject>;
export type DataSources = {
  transactionsDb: TransactionsDb;
};

export type Context = {
  dataSources: DataSources;
};

export const createDataSources = (mongoClient: MongoClient): DataSources => {
  const collection = (name: string) => mongoClient.db().collection(name);

  return {
    transactionsDb: new MongoDataSource<TransactionDbObject>(collection('transactions')),
  };
};

export const createApolloServer = (dataSources: DataSources): ApolloServer => {
  const resolvers = loadFilesSync(join(__dirname, './**/*-resolvers.*'));
  const schema = makeExecutableSchema({
    resolvers,
    typeDefs: [typeDefs, typeDefsMongo],
    inheritResolversFromInterfaces: true,
    resolverValidationOptions: {
      requireResolversForResolveType: false,
    },
  });
  return new ApolloServer({ schema, dataSources: () => dataSources });
};
