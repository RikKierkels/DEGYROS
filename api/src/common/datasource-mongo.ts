import { DataSource, DataSourceConfig } from 'apollo-datasource';
import { InMemoryLRUCache, KeyValueCache } from 'apollo-server-caching';
import { Collection, FilterQuery, ObjectId } from 'mongodb';
import DataLoader from 'dataloader';
import { hasValue } from './util';
import { EJSON } from 'bson';
import { Context } from '../apollo';

type MongoDataSourceConfig = {
  timeToLive: number;
};

const MINUTE_IN_MS = 60000;
const DEFAULT_CONFIG: MongoDataSourceConfig = {
  timeToLive: MINUTE_IN_MS,
};

export class MongoDataSource<T extends { _id: ObjectId }> extends DataSource {
  private cache: KeyValueCache = new InMemoryLRUCache();
  private readonly config: MongoDataSourceConfig;
  private loader: DataLoader<string, T>;

  public collection: Collection<T>;

  constructor(collection: Collection<T>, config: MongoDataSourceConfig = DEFAULT_CONFIG) {
    super();
    this.config = config;
    this.loader = this.createLoader();
    this.collection = collection;
  }

  initialize({ cache = this.cache }: DataSourceConfig<any>): void {
    this.cache = cache;
  }

  // TODO: Get rid of cast to FilterQuery
  createLoader(): DataLoader<string, T> {
    return new DataLoader<string, T>((ids) => {
      return this.collection
        .find({ _id: { $in: ids } } as FilterQuery<T>)
        .toArray()
        .then(this.preserveLoaderOrder(ids));
    });
  }

  private preserveLoaderOrder(ids: readonly string[]): (documents: T[]) => T[] {
    return (documents) => {
      return ids.map((id) => documents.find(({ _id }) => _id.equals(id))).filter(hasValue);
    };
  }

  async findOneById(id: string): Promise<T> {
    const key = this.createCacheKey(id);

    const documentFromCache = await this.cache.get(key);
    if (documentFromCache) {
      return EJSON.parse(documentFromCache) as T;
    }

    const document = await this.loader.load(id);
    const { timeToLive } = this.config;
    if (timeToLive) {
      this.cache.set(key, EJSON.stringify(document), { ttl: timeToLive });
    }

    return document;
  }

  async findManyById(ids: string[]): Promise<T[]> {
    return Promise.all(ids.map(this.findOneById));
  }

  async removeFromCacheById(id: string): Promise<void> {
    const key = this.createCacheKey(id);
    this.loader.clear(id);
    await this.cache.delete(key);
  }

  private createCacheKey(id: string): string {
    return `mongo-${this.collection.collectionName}-${id}`;
  }
}
