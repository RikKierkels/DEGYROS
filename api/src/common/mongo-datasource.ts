import { DataSource, DataSourceConfig } from 'apollo-datasource';
import { InMemoryLRUCache, KeyValueCache } from 'apollo-server-caching';
import { Collection, FilterQuery, ObjectId } from 'mongodb';
import DataLoader from 'dataloader';
import { EJSON } from 'bson';
import { Context } from '../apollo';

type Id = ObjectId | string;
type MongoDataSourceConfig = {
  ttl: number;
};

const MINUTE_IN_MS = 60000;
const DEFAULT_CONFIG: MongoDataSourceConfig = {
  ttl: MINUTE_IN_MS,
};

export class MongoDataSource<T extends { _id: Id }> extends DataSource {
  private cache: KeyValueCache = new InMemoryLRUCache();
  private readonly config: MongoDataSourceConfig;
  private loader: DataLoader<Id, T>;

  public collection: Collection<T>;

  constructor(collection: Collection<T>, config: MongoDataSourceConfig = DEFAULT_CONFIG) {
    super();
    this.config = config;
    this.loader = this.createLoader();
    this.collection = collection;
  }

  initialize({ cache = this.cache }: DataSourceConfig<Context>): void {
    this.cache = cache;
  }

  createLoader(): DataLoader<Id, T> {
    return new DataLoader((ids) => {
      return this.collection
        .find({ _id: { $in: ids } } as FilterQuery<T>)
        .toArray()
        .then(this.preserveOriginalOrder(ids));
    });
  }

  private preserveOriginalOrder(ids: readonly Id[]): (documents: T[]) => T[] {
    return (documents) => {
      return ids.map((id) => {
        const index = documents.findIndex(({ _id }) => this.idToString(_id) === this.idToString(id));
        return documents[index];
      });
    };
  }

  private idToString(id: Id): string {
    return id instanceof ObjectId ? id.toHexString() : id;
  }

  private createCacheKey(id: Id): string {
    return `mongo-${this.collection.collectionName}-${this.idToString(id)}`;
  }

  async findOneById(id: Id): Promise<T> {
    const key = this.createCacheKey(id);

    const documentFromCache = await this.cache.get(key);
    if (documentFromCache) {
      return EJSON.parse(documentFromCache) as T;
    }

    const document = await this.loader.load(id);
    const { ttl } = this.config;
    if (document && ttl) {
      this.cache.set(key, EJSON.stringify(document), { ttl });
    }

    return document;
  }

  async findManyById(ids: Id[]): Promise<T[]> {
    const documents = ids.map((id) => this.findOneById(id));
    return Promise.all(documents).then((documents) => documents.filter((doc) => doc));
  }

  async removeFromCacheById(id: Id): Promise<void> {
    const key = this.createCacheKey(id);
    this.loader.clear(id);
    await this.cache.delete(key);
  }
}
