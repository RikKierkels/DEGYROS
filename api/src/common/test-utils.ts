import { createTestClient } from 'apollo-server-testing';
import { MongoMemoryServer } from 'mongodb-memory-server';
import { createApolloServer, DataSources } from '../apollo';
import { createMongoClient } from '../mongo';
import { MongoClient } from 'mongodb';

export type MongoMemoryClient = {
  instance: () => MongoClient;
  stop: () => Promise<boolean>;
};

export const createMongoMemoryClient = async (): Promise<MongoMemoryClient> => {
  const mongoServer = new MongoMemoryServer();
  const uri = await mongoServer.getUri();
  const client = await createMongoClient(uri);

  return {
    instance: () => client,
    stop: () => (client.close(), mongoServer.stop()),
  };
};

export const createApolloTestClient = (dataSources: DataSources) => {
  const server = createApolloServer(dataSources);
  return createTestClient(server as any);
};
