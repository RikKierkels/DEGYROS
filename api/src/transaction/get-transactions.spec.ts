import { gql } from 'apollo-server';
import { createApolloTestClient, createMongoMemoryClient, MongoMemoryClient } from '../common/test-utils';
import { createDataSources } from '../apollo';

let mongoClient: MongoMemoryClient;

beforeEach(async () => {
  mongoClient = await createMongoMemoryClient();
});

afterEach(async () => {
  return mongoClient.stop();
});

test('given existing transactions, when retrieving transactions, returns all transactions', async () => {
  const dataSources = createDataSources(mongoClient.instance());
  const { query } = createApolloTestClient(dataSources);

  await dataSources.transaction.collection.insertMany([
    {
      _id: '03065814-a998-4876-8b6a-bafdeb26664e',
      purchaseDate: '2020-11-12T21:55:00.000Z',
      product: 'HSBC MSCI WORLD',
      ISIN: 'IE00B4X9L533',
      exchange: 'EPA',
      count: 2,
      rate: {
        amount: 20339,
        numberOfDecimals: 3,
        currency: 'EUR',
      },
      purchaseValue: {
        amount: 4068,
        numberOfDecimals: 2,
        currency: 'EUR',
      },
      costs: {
        amount: 0,
        numberOfDecimals: 2,
        currency: 'EUR',
      },
      total: {
        amount: 4068,
        numberOfDecimals: 2,
        currency: 'EUR',
      },
    },
    {
      _id: '29f09ed1-684c-4df3-8295-e12b7e8460d6',
      purchaseDate: '2020-12-31T09:55:00.000Z',
      product: 'VANGUARD S&P500',
      ISIN: 'IE00B3XXRP09',
      exchange: 'EAM',
      count: 2,
      rate: {
        amount: 54981,
        numberOfDecimals: 3,
        currency: 'EUR',
      },
      purchaseValue: {
        amount: 10996,
        numberOfDecimals: 2,
        currency: 'EUR',
      },
      costs: {
        amount: 0,
        numberOfDecimals: 2,
        currency: 'EUR',
      },
      total: {
        amount: 10996,
        numberOfDecimals: 2,
        currency: 'EUR',
      },
    },
  ]);

  const transactionsQuery = gql`
    fragment priceFields on Price {
      amount
      numberOfDecimals
      currency
    }

    query transactions {
      transactions {
        id
        purchaseDate
        product
        ISIN
        exchange
        count
        rate {
          ...priceFields
        }
        purchaseValue {
          ...priceFields
        }
        costs {
          ...priceFields
        }
        total {
          ...priceFields
        }
      }
    }
  `;

  const { data, errors } = await query({ query: transactionsQuery });

  expect(errors).toBeUndefined();
  expect(data).toMatchSnapshot();
});
