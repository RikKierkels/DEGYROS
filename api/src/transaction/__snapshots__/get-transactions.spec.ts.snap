// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given existing transactions, when retrieving transactions, returns all transactions 1`] = `
Object {
  "transactions": Array [
    Object {
      "ISIN": "IE00B4X9L533",
      "costs": Object {
        "amount": 0,
        "currency": "EUR",
        "numberOfDecimals": 2,
      },
      "count": 2,
      "exchange": "EPA",
      "id": "03065814-a998-4876-8b6a-bafdeb26664e",
      "product": "HSBC MSCI WORLD",
      "purchaseDate": 2020-11-12T21:55:00.000Z,
      "purchaseValue": Object {
        "amount": 4068,
        "currency": "EUR",
        "numberOfDecimals": 2,
      },
      "rate": Object {
        "amount": 20339,
        "currency": "EUR",
        "numberOfDecimals": 3,
      },
      "total": Object {
        "amount": 4068,
        "currency": "EUR",
        "numberOfDecimals": 2,
      },
    },
    Object {
      "ISIN": "IE00B3XXRP09",
      "costs": Object {
        "amount": 0,
        "currency": "EUR",
        "numberOfDecimals": 2,
      },
      "count": 2,
      "exchange": "EAM",
      "id": "29f09ed1-684c-4df3-8295-e12b7e8460d6",
      "product": "VANGUARD S&P500",
      "purchaseDate": 2020-12-31T09:55:00.000Z,
      "purchaseValue": Object {
        "amount": 10996,
        "currency": "EUR",
        "numberOfDecimals": 2,
      },
      "rate": Object {
        "amount": 54981,
        "currency": "EUR",
        "numberOfDecimals": 3,
      },
      "total": Object {
        "amount": 10996,
        "currency": "EUR",
        "numberOfDecimals": 2,
      },
    },
  ],
}
`;
